File: src/j/implementation/microSet/InnerSet.java
package j.implementation.microSet;

import java.util.Collection;
import java.util.Iterator;

public interface InnerSet<T> {

    InnerSet<T> addElement(T p_element);

    InnerSet<T> addUnChecked(T p_element);

    InnerSet<T> addAllElements(InnerSet<T> innerSet);

    boolean containsElement(Object p_element);

    boolean containsAllElements(InnerSet<T> innerSet);

    T getElement(int index);

    InnerSet<T> removeElement(Object p_element);

    InnerSet<T> removeAllElements(InnerSet<T> innerSet);

    InnerSet<T> retainAllElements(InnerSet<T> innerSet);

    InnerSet<T> addAndPropagate(T p_element, MicroSet<T> microSet);

    InnerSet<T> addAllAndPropagate(InnerSet<T> innerSet, MicroSet<T> microSet);

    InnerSet<T> addAllAndPropagateReverse(InnerSet<T> innerSet, MicroSet<T> microSet);

    int getSize();

    Iterator<T> iterator();

    InnerSet<T> copy();

    InnerSet<T> clear(boolean unused);

}

File: src/scala/implementation/InnerSet.scala
package scala.implementation

trait InnerSet[T] {

  def addElement(p_element: T): InnerSet[T]

  def addUnchecked(p_element: T): InnerSet[T]

  def addAndPropagate(p_element: T, p_microSet: MicroSet[T]): InnerSet[T]

  def addAllElements(p_innerSet: InnerSet[T]): InnerSet[T]

  def addAllAndPropagate(p_innerSet: InnerSet[T],p_microSet: MicroSet[T]): InnerSet[T]

  def containsElement(p_element: Object): Boolean

  def containsAllElements(p_innerSet: InnerSet[T]): Boolean

  def getElement(index: Int): Option[T]

  def removeElement(p_element: Object): InnerSet[T]

  def removeAllElements(p_innerSet: InnerSet[T]): InnerSet[T]

  def retainAllElements(p_innerSet: InnerSet[T]): InnerSet[T]

  def getSize: Int

  def iterator: InnerSetIterator[T]

  def copy: Option[InnerSet[T]]

  def clear(unused: Boolean): InnerSet[T]

  def toString: String
}
